cdap-common/src/main/java/io/cdap/cdap/common/internal/remote/RemoteClient.java

```diff
@@ -108,17 +110,26 @@ public HttpResponse execute(HttpRequest request) throws IOException, Unauthorize
 
     // Add Authorization header and use a rewritten URL if needed
     RemoteAuthenticator authenticator = getAuthenticator();
-    if (authenticator != null) {
-      if (headers.keySet().stream().noneMatch(HttpHeaders.AUTHORIZATION::equalsIgnoreCase)) {
-        headers.put(HttpHeaders.AUTHORIZATION,
-                    String.format("%s %s", authenticator.getType(), authenticator.getCredentials()));
-      }
+    if (authenticator != null && headers.keySet().stream().noneMatch(HttpHeaders.AUTHORIZATION::equalsIgnoreCase)) {
+      headers.put(HttpHeaders.AUTHORIZATION, String.format("%s %s", authenticator.getType(),
+                                                           authenticator.getCredentials()));
     }
 
-    String internalCredentials = authenticationContext.getPrincipal().getCredential();
-    if (internalCredentials != null) {
-      //TODO CDAP-17772 - add credentials type
-      headers.put(Constants.Security.Headers.RUNTIME_TOKEN, internalCredentials);
+    Principal principal = authenticationContext.getPrincipal();
+    String userID = null;
+    Credential internalCredentials = null;
+    if (principal != null) {
+      userID = principal.getName();
+      internalCredentials = principal.getCredential();
+    }
+    if (internalCredentials != null &&
+      headers.keySet().stream().noneMatch(Constants.Security.Headers.RUNTIME_TOKEN::equalsIgnoreCase)) {
+      headers.put(Constants.Security.Headers.RUNTIME_TOKEN,
+                  String.format("%s %s", internalCredentials.getType().getQualifiedName(),
+                                internalCredentials.getValue()));
+    }
+    if (userID != null && headers.keySet().stream().noneMatch(Constants.Security.Headers.USER_ID::equalsIgnoreCase)) {
```

Comment:
Same here

Reply:
Also I'd go "all or none" principle. It's weird to set just some headers and propagate others and can potentially be a security issue

Reply:
I agree, went ahead and refactored all the existing locations in which the user ID header is preset.